modal/file_io.py

```diff
@@ -0,0 +1,403 @@
+# Copyright Modal Labs 2024
+import asyncio
+import io
+from typing import AsyncIterator, Generic, Literal, Optional, Sequence, TypeVar, Union, cast
+
+from grpclib.exceptions import GRPCError, StreamTerminatedError
+
+from modal._utils.grpc_utils import retry_transient_errors
+from modal_proto import api_pb2
+
+from ._utils.async_utils import synchronize_api
+from ._utils.grpc_utils import RETRYABLE_GRPC_STATUS_CODES
+from .client import _Client
+from .exception import FilesystemExecutionError
+
+OpenTextModeUpdating = Literal["r+", "+r", "w+", "+w", "a+", "+a", "x+", "+x"]
+OpenTextModeWriting = Literal["w", "a", "x"]
+OpenTextModeReading = Literal["r"]
+OpenTextMode = Union[OpenTextModeUpdating, OpenTextModeWriting, OpenTextModeReading]
+OpenBinaryModeUpdating = Literal[
```

Comment:
I think there's gotta be a better way to do this than listing out every ordering. Can you not just check that the characters are in the list?

Reply:
I agree that listing all these out isn't ideal. Type safety is built into Python's native `open`, which was my original incentive to add it here. 

If we want to preserve this, another option is to take these variables from [_typeshed](https://github.com/python/typeshed/tree/main/stdlib/_typeshed). This would also need some changes to synchronicity to get working: https://github.com/modal-labs/synchronicity/pull/193. I don't think these changes are compatible with python 3.8 / I'm not sure if the changes in that draft pr would necessarily be the correct way to approach resolving `_typeshed`.

I'm ok with removing the type safety is this is too much, though. Deferring to @mwaskom / @freider.